minifiers/config.js

```diff
@@ -0,0 +1,148 @@
+  if (config.mangle && config.mangle.properties && config.mangle.properties.regex) {
+    config.mangle.properties.regex = regexp(config.mangle.properties.regex);
+  }
+
+  return config;
+}
+
+/**
+ * Normalized configuration for various minifiers.
+ *
+ * @class Config
+ * @param {Object} config Minification configuration from wrhs.toml.
+ * @param {String} filename Reference to minified filename.
+ * @param {String} map Reference to sourcemap url.
+ * @public
+ */
+class Config {
+  constructor({ minify, sourceMapContent, filename } = {}) {
+    this.filename = filename;
+
+    //
+    // Prevent polution of wrhs.toml by normalizing and cloning configuration.
+    //
+    this.values = this.normalize(minify, sourceMapContent);
+  }
+
+  /**
+   * Normalize configuration to have required set of options.
+   *
+   * @param {Object} minify base configuration from wrhs.toml.
+   * @param {String} sourceMapContent JSON representation of sourceMap.
+   * @returns {Object} configuration
+   * @private
+   */
+  normalize(minify = {}, sourceMapContent) {
+    const config = rip(cloneDeep(minify), 'minifier');
+
+    config.parse = config.parse || {};
+    config.compress = config.compress || {};
+    config.sourceMap = config.sourceMap || {};
+
+    //
+    // Define sourcemap config with:
+    //  - filename: reference to minified file the sourcemap maps to
+    //  - url: sourcemap url to append to code content
+    //
+    config.sourceMap.includeSources = true;
+    config.sourceMap.filename = this.filename;
+    config.sourceMap.url = this.map;
+
+    //
+    // Minifier was given sourcemap content as input from earlier minification.
+    // Test if the content is valid JSON and add it to the content.
+    //
+    if (tryParse(sourceMapContent)) {
+      config.sourceMap.content = sourceMapContent.toString('utf-8');
+    }
+
+    return config;
+  }
+
+  /**
+   * Return filename of the sourceMap.
+   *
+   * @returns {String} filename
+   * @public
+   */
+  get map() {
+    return `${this.filename}.map`;
+  }
+
+  /**
+   * Return Terser configuration. Clone to prevent downstream changes to `values`.
+   *
+   * @returns {Object} configuration
+   * @public
+   */
+  get terser() {
+    return transformMangleRegex(
+      cloneDeep(this.values)
+    );
+  }
+
+  /**
+   * Return UglifyJS configuration. Clone to prevent downstream changes to `values`.
+   *
+   * @returns {Object} configuration
+   * @public
+   */
+  get uglifyjs() {
+    const { mangleProperties } = this.values;
+    let config = cloneDeep(
+      rip(this.values, 'mangleProperties')
+    );
+
+    config = transformMangleRegex(config);
+
+    //
+    // Support legacy uglify-js option and transform it to support uglify-js@3
+    //
+    if (typeof mangleProperties === 'object') {
+      if (typeof config.mangle !== 'object') config.mangle = {};
+      if (mangleProperties.regex) mangleProperties.regex = regexp(mangleProperties.regex);
```

Comment:
You did a lot of work with `cloneDeep` to not touch the original `values` object, but this is touching it, are you sure you want to do that?

Reply:
Good call, must have been an accident move after fixing some of the monkey patching from uglifyjs2

Reply:
Sealed the deal by freezing the normalized config object :)

Reply:
Sealed the deal by freezing the normalized config object :)